float64 weight
float64[] centroid
float64[] precision_matrix